### PR TITLE
feat: add unused HPA detection

### PR DESCRIPTION
**Pull request by shay-legit**

### CODE REVIEW

File: pkg/kor/hpas.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+	for _, hpa := range hpas {
+		if hpa.TargetKind == "Deployment" {
+			if !slices.Contains(deploymentNames, hpa.TargetName) {
+				diff = append(diff, hpa.Name)
+			}
+		} else if hpa.TargetKind == "StatefulSet" {
+			if !slices.Contains(statefulsetNames, hpa.TargetName) {
+				diff = append(diff, hpa.Name)
+			}
+		}
```
Comment (luisdavim) [inline pkg/kor/hpas.go:87]: nit: use swtich/case

**Comment by yonahd** [comment]
The HPA cmd should be supported in the `all` command as well.

File: pkg/kor/hpas.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+	}
+	hpas := make([]Hpa, 0, len(rawHpas.Items))
+	for _, hpa := range rawHpas.Items {
+		hpas = append(hpas, Hpa{
+			TargetKind: hpa.Spec.ScaleTargetRef.Kind,
+			TargetName: hpa.Spec.ScaleTargetRef.Name,
+			Name:       hpa.Name,
+			Namespace:  hpa.Namespace,
+		})
+	}
```
Comment (luisdavim) [inline pkg/kor/hpas.go:59]: is the custom `Hpa` type really needed? why not just iterate though the hpa list we get from the API? This way 2 loops are needed.

File: pkg/kor/hpas.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+		} else if hpa.TargetKind == "StatefulSet" {
+			if !slices.Contains(statefulsetNames, hpa.TargetName) {
+				diff = append(diff, hpa.Name)
+			}
+		}
+	}
+	return diff, nil
+}
+
+func ProcessNamespaceHpas(clientset *kubernetes.Clientset, namespace string) ([]string, error) {
```
Comment (luisdavim) [inline pkg/kor/hpas.go:92]: does this method need to be public? I don't see it being used outside of this package.

File: cmd/kor/hpas.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+		}
+	},
+}
+
+func init() {
+	hpaCmd.PersistentFlags().StringVarP(&kubeconfig, "kubeconfig", "k", "", "Path to kubeconfig file (optional)")
+	hpaCmd.PersistentFlags().StringVarP(&namespace, "namespace", "n", "", "Namespace to run on")
+	hpaCmd.PersistentFlags().StringVar(&outputFormat, "output", "table", "Output format (table or json)")
+	rootCmd.AddCommand(hpaCmd)
+}
```
Comment (luisdavim) [inline cmd/kor/hpas.go:26]: these flags seem to be getting set on every command, maybe they should be set at the root as `PersistentFlags` so they are available on every child command.

**Comment by shay-legit** [comment]
> The HPA cmd should be supported in the `all` command as well.

good catch! works nicely.

image

File: cmd/kor/hpas.go (hunk 4 above)
Comment (shay-legit) [reply cmd/kor/hpas.go:26]: honestly i followed the code base - didn't wanna draft too much from the other resources.
WDYT @yonahd  ?

Comment (yonahd) [reply cmd/kor/hpas.go:26]: You can leave it this way now. I'll do the refactor across the code base.

File: pkg/kor/hpas.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+		diff, err := processNamespaceHpas(kubeClient, namespace)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Failed to process namespace %s: %v\n", namespace, err)
+			continue
+		}
+		if len(diff) > 0 {
+			if response[namespace] == nil {
+				response[namespace] = make(map[string][]string)
+			}
+			response[namespace]["Hpas"] = diff
```
Comment (yonahd) [inline pkg/kor/hpas.go:141]: Can you make this singular
`response[namespace]["Hpa"] = diff`

File: pkg/kor/all.go [hunk 6]
```diff
@@ -80,6 +80,15 @@ func getUnusedRoles(kubeClient *kubernetes.Clientset, namespace string) Resource
 	return namespaceSADiff
 }
 
+func getUnusedHpas(kubeClient *kubernetes.Clientset, namespace string) ResourceDiff {
+	hpaDiff, err := processNamespaceHpas(kubeClient, namespace)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to get %s namespace %s: %v\n", "hpas", namespace, err)
+	}
+	namespaceSADiff := ResourceDiff{"HPA", hpaDiff}
```
Comment (yonahd) [inline pkg/kor/all.go:88]: Change to lowercase to match the return from the singular json function
namespaceSADiff := ResourceDiff{"Hpa", hpaDiff}

File: pkg/kor/hpas.go (hunk 2 above)
Comment (shay-legit) [reply pkg/kor/hpas.go:59]: not a must but if u think it's better without i will change it, no problem

Comment (shay-legit) [reply pkg/kor/hpas.go:59]: done

**Comment by yonahd** [comment]
Looking good!
Can you just sort your imports using 
`goimports -w .`

**Review by yonahd — APPROVED** [review]
LGTM
Thanks for the contribution!

**Pull request merged**